CommunicationLayer/CommDeviceControl/UdpMessageForwarder.cpp

```
@@ -26,27 +26,69 @@
 #include <QByteArray>
 #include <QDebug>
 #include <QHostAddress>
+#include <QTextCodec>
+
 
 #include "InfrastructureLayer/Settings/I_Settings.h"
 #include "UdpMessageForwarder.h"
 
+
+namespace {
+    quint32 TIMEOUT = 120;
+    quint32 SLEEP_TIME = 2;
+}
+
 UdpMessageForwarder::UdpMessageForwarder(I_Settings& settings)
-: ipAddress_(settings.ipAddress())
-, port_(settings.udpPort())
 {
-    socket_.bind(ipAddress_, port_);
+    exchangeName_ = settings.exchangeName();
+    ipAddress_ = settings.ipAddress();
+    udpPort_ = settings.udpPort();
+    setupChannel();
 }
 
 UdpMessageForwarder::~UdpMessageForwarder()
 {
 }
 
+
+void UdpMessageForwarder::setupChannel()
+{
+    quint32 i = 0;
+    do {
+        if (i++) {
+            qWarning() << "UdpMessageForwarder: Attempting to reconnect";
+        }
+        try {
+            channel_ = AmqpClient::Channel::Create(ipAddress_.toStdString(), udpPort_);
+        } catch (std::exception&) {
+            if (channel_ == NULL) {
+                if (i == (TIMEOUT / SLEEP_TIME)) {
+                    qWarning() << "UdpMessageForwarder timed out waiting for connection to broker";
+                    throw;
+                }
+                qWarning() << "UdpMessageForwarder: error creating channel";
+                QThread::sleep(SLEEP_TIME);
+            }
+            else {
+                throw;
+            }
+        }
+
+
+    } while (channel_ == NULL);
+    channel_->DeclareExchange(exchangeName_.toStdString(), AmqpClient::Channel::EXCHANGE_TYPE_FANOUT);
+}
```

@enochtsang So my idea here is to implement Hermes trying to reconnect to the rabbitmq broker instead of just terminating if it can't find it, I can add the timeout and sleep time to the settings if you like it. Unfortunately I can't seem to catch one of the AmqpExceptions here as I did elsewhere because it's thrown from boost if this fails, and seems to be a generic C++ exception. Let me know your opinion

---

I think the program should terminate. In our use case, we can't really turn the server on after the car turns on, so failing fast and having hermes erroring out makes sense I think.

---

But what if the server is running and there is just some temporary connectivity issue?

---

I think the big problem here is that, since hermes is a background process, we won't really have a way to indicate if some problem occurred. I don't think it makes sense in the program to create any pop ups, but if we can close the program with a bad exit code, then we can delegate that responsibility to another program (Domovoi).